File: 32. Longest Valid Parentheses/32.cpp
//改自https://books.halfrost.com/leetcode/ChapterFour/0001~0099/0032.Longest-Valid-Parentheses/
class Solution
{
public:
    int longestValidParentheses(string s)
    {
        stack<int> stack;
        stack.push(-1); //初始化-1让左边有个起点，当第一个为')'时，可以抵掉
        int res = 0;
        for (int i = 0; i < s.size(); i++)
        {
            if (s[i] == '(') //遇到'('，压栈
            {
                stack.push(i);
            }
            else //遇到')'，弹栈
            {
                stack.pop();
                if (stack.empty()) //如果栈空，说明这个')'是无效的，记录下起点
                {
                    stack.push(i);
                }
                else //如果栈不空，说明这个')'是有效的，计算长度
                {
                    res = max(res, i - stack.top());
                }
            }
        }
        return res;
    }
};

File: 32. Longest Valid Parentheses/32.py
class Solution:

    def longestValidParentheses(self, s: str) -> int:
        stack, res = [], 0
        stack.append(-1)
        for i in range(len(s)):
            if s[i] == '(':
                stack.append(i)
            else:
                stack.pop()
                if not stack:
                    stack.append(i)
                else:
                    res = max(res, i - stack[-1])
        return res
